### PR TITLE
Default compilation pipelines

### PR DESCRIPTION
**Pull request by NeuralCoder3**
We want to avoid having the user define a whole pipeline in every compiled file.
In most cases, the pipeline should only depend on the loaded/involved dialects.

Assuming that every dialect provides a pipeline part (phase/pass) that can be composed, 
only the composition has to happen depending on whether a dialect is loaded.

I see three (non-exclusive) possibilities to accomplish this goal:

### Includes
A general import allows for effective code sharing between files.
The user would define the pipeline once and include it in all of the files.

```rust
.include "./ad_pipeline.thorin";
...
``` 

### Normalizers (better from a theoretical point of view)
```rust
.let _compile [] = {
%compile.pipe
   (dummy_phase, demo_phase)#(is_loaded %opt.demo_dialect)
}
``` 

We would need a way to register dialects in a global storage that is accessible during normalization.
As all dialects are registered before the normalization fires, we do not have to take care of the normalization timestamp.

```rust
.let _compile [dialects:%opt.dialects] = {
%compile.pipe
   (dummy_phase, demo_phase)#(contains_dialect dialects %opt.demo_dialect)
}
``` 
Using this variant, we could shift the access burdon a bit.
If `dialects` can forward the access, the normalizer does not have to look the dialects up from the C++ side.
However, this might just redistribute the problem.

### Passes (easier to implement)
```rust
.let _compile [] = {
%compile.pipe
   (%opt.dialect_phase %opt.demo_dialect demo_phase dummy_phase)
  //                                   ^ dialect to test        ^ then phase ^ else phase
}
``` 
Instead of handling dialects in the normalizer, we can handle them in the pass executors.
Using this method, we would need to give the dialects access to a registering mechanism in the pass-registering method and the check mechanism in the pass-building process.

### Axioms

We could also inspect the world for loaded axioms. But this would not necessarily be the same as loaded plugins.
In particular, the opt dialect would load all dialects to access their passes.

### CODE REVIEW

**Comment by NeuralCoder3** [comment]
I chose the passes approach as it can be implemented in an unintrusive manner.
The new opt dialect loads all thorin files to access the passes.
Depending on the loaded dialects, the correct passes are assembled into a default compilation pipeline.
Currently, I set the opt dialect to be always loaded implicitly (along with `mem`, `core`, and `compile`).
If it is not loaded, the compilation could still fall through to the minimal compilation from `compile`.
We could also introduce an intermediate compilation fallback in `mem` as it is always loaded.

Note: One complication is that internals are currently not natively supported, and thus, the `autodiff` internals are simulated using externals. Therefore, an internal cleanup phase is necessary if `autodiff` is not loaded (and is currently performed in every pipeline).

**Comment by NeuralCoder3** [comment]
For documentation purposes:

<details>
  <summary>Old Pipeline</summary>
``` 
1 Scalerize
2 EtaRed
3 TailRecElim

100 
  Opt
  Affine
    LowerFor
  Mem
   BetaRed
   EtaRed
   EtaExp er
   SSA ee
   CopyProp br ee false

AD
  105 ADEval
  110 OPT
  111 ADZero

Direct
  115 DS2CPS
  116 CPS2DS
  120 OPT

Clos
  121 ClosConvPrep nullptr
  ++  EtaExp nullptr
  ++  ClosConvWrapper
  ++  
    EtaRed true
    EtaExp er
    Scalerize ee
  ++ 
    Scalerize nullptr
    BranchElim
    CopyProp nullptr nullptr true
    LowerTypedClosPrep
    Clos2SJLJ
  ++  LowerTypedClosWrapper

200 LamSpec (controversial)

AD
 299 
   ADZeroCleanup
   ADExternalCleanup

300 CodeGenProp 
  RetWrap
  Mem
    RememElim
    Alloc2Malloc
  Refly
    RemoveDbgPerm
``` 
</details>

**Comment by leissa** [comment]
Well, I can't follow here exactly. I guess we can discuss this in our next online meeting. But so far it looks pretty cool :)
But before we overdesign stuff, can't we just have one pipeline that is maintained by us that includes all Passes/Phases from dialects that are maintained by us. Now, for the 3rd party plugin you would need to write your own pipeline, but so what?

**Comment by NeuralCoder3** [comment]
> But before we overdesign stuff, can't we just have one pipeline that is maintained by us that includes all Passes/Phases from dialects that are maintained by us. 

This is what is currently implemented with the minimum amount of overdesigning :) :
https://github.com/NeuralCoder3/thorin2/blob/default-compilation/dialects/opt/opt.thorin#L66

The problem is that we need to know what passes are available / what dialects are loaded to compose the pipeline correctly in the presence of dialects.

Now thinking about it, we could just throw all passes together and at pipeline-building blindly try to test whether the requested pass exists or not. This would be even simpler but would also complicate debugging/development compared to the current approach.

**Comment by leissa** [comment]
So, the problem is that for some test cases we basically have disabled some passes/phases - simply because we haven't loaded the appropriate plugin?

**Comment by NeuralCoder3** [comment]
Yes, but also the inverse holds:
If we do not need a dialect, we probably also do not want it to run all of its passes.

Also, some passes introduce unnecessary overhead if not used.
For instance, `autodiff` rewrites the whole program multiple times (reshape before, reshape + closure after; and two cleanup phases in the end).

**Comment by NeuralCoder3** [comment]
As comment: [df9eb5f](https://github.com/AnyDSL/thorin2/pull/154/commits/df9eb5f28f642832a462461545724ac06f4dce41) removes the additional layers of indirection in the pass building process. This also solves the issue that passes only received nullptr as pass arguments in the current pipeline.

**Comment by leissa** [comment]
I guess the compile dialect is missing somewhere as dependency - hence the doxygen fail.

File: docs/Doxyfile.in [hunk 1]
```diff
@@ -866,7 +866,7 @@ WARN_NO_PARAMDOC       = NO
 # Possible values are: NO, YES and FAIL_ON_WARNINGS.
 # The default value is: NO.
 
-WARN_AS_ERROR          = YES
```
Comment (leissa) [inline docs/Doxyfile.in:869]: Can we turn this on again?

File: thorin/pass/pipelinebuilder.cpp [hunk 2]
```diff
@@ -1,19 +1,23 @@
 #include "thorin/pass/pipelinebuilder.h"
 
-#include <compare>
+// #include <compare>
 
-#include <vector>
+// #include <memory>
+// #include <vector>
```
Comment (leissa) [inline thorin/pass/pipelinebuilder.cpp:6]: remove

File: thorin/pass/pipelinebuilder.h [hunk 3]
```diff
@@ -1,74 +1,53 @@
...
 
 class PipelineBuilder {
 public:
-    explicit PipelineBuilder() {}
-
-    int last_phase();
+    PipelineBuilder(World& world)
+        : world_(world) {
+        pipe = std::make_unique<Pipeline>(world);
+        man  = nullptr;
```
Comment (leissa) [inline thorin/pass/pipelinebuilder.h:18]: not needed. std::unique_ptr will be initalized with `nullptr`

File: thorin/pass/pipelinebuilder.h [hunk 4]
```diff
@@ -1,74 +1,53 @@
...
+using PassInstanceMap = absl::btree_map<const Def*, Pass*, GIDLt<const Def*>>;
 
 class PipelineBuilder {
 public:
-    explicit PipelineBuilder() {}
-
-    int last_phase();
+    PipelineBuilder(World& world)
+        : world_(world) {
+        pipe = std::make_unique<Pipeline>(world);
```
Comment (leissa) [inline thorin/pass/pipelinebuilder.h:17]: better
```c++
    : world_(world)
    , pipe(std::make_unique<Pipeline>(world) {
```

**Pull request merged**